### PR TITLE
Add listVaults rpc

### PR DESCRIPTION
**Pull request by jingyi2811**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

Add listVaults rpc

#### Which issue(s) does this PR fixes?:

Fixes part of #488

### CODE REVIEW

File: packages/jellyfish-api-core/src/category/loan.ts [hunk 1]
```diff
@@ -204,6 +204,15 @@ export class Loan {
     return await this.client.call('getvault', [vaultId], 'bignumber')
   }
 
+  /**
+   * List all available vaults.
+   *
+   * @return {Promise<VaultData>}
```
Comment (surangap) [inline packages/jellyfish-api-core/src/category/loan.ts:210]:
```suggestion
   * @return {Promise<Record<string, VaultDetails>>} Object including details of the vaults.
```

**Comment by jingyi2811** [comment]
Approved + 1

**Review by canonbrother — DISMISSED** [review]

File: packages/jellyfish-api-core/src/category/loan.ts [hunk 2]
```diff
@@ -204,6 +204,15 @@ export class Loan {
     return await this.client.call('getvault', [vaultId], 'bignumber')
   }
 
+  /**
+   * List all available vaults.
+   *
+   * @return {Promise<Record<string, VaultDetails>>} Object including details of the vaults.
+   */
+  async listVaults (): Promise<Record<string, VaultDetails>> {
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/src/category/loan.ts:212]: Currently, there isn't a way to listVaults with pagination or filtering?

Comment (surangap) [reply packages/jellyfish-api-core/src/category/loan.ts:212]: yes. there's no filtering or pagination for this RPC atm.

Comment (fuxingloh) [reply packages/jellyfish-api-core/src/category/loan.ts:212]: Damn, this is gonna be awkward. We need indexed address and pagination for Light Wallet.

File: docs/node/RPC Category/16-loan.md [hunk 3]
```diff
@@ -330,6 +330,48 @@ interface loan {
 }
 
 interface VaultDetails {
+  vaultId?: string
```
Comment (surangap) [inline docs/node/RPC Category/16-loan.md:333]: @Jouzo `vaultId` is currently not returned by `getvault` RPC. should we add it there? one have to know the `vaultId` to call `getvault`. But should we add it for consistency?

File: packages/jellyfish-api-core/__tests__/category/loan/listVaults.test.ts [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+        vaultId: vaultId3,
+        ownerAddress: ownerAddress3,
+        loanSchemeId: 'scheme',
+        isUnderLiquidation: false
+      }
+    ]))
+  })
+})
+
+describe('Loan listVault with options and pagination', () => {
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/loan/listVaults.test.ts:130]:
```suggestion
describe('Loan listVaults with options and pagination', () => {
```

File: packages/jellyfish-api-core/__tests__/category/loan/listVaults.test.ts [hunk 5]
```diff
@@ -0,0 +1,386 @@
+import { LoanMasterNodeRegTestContainer } from './loan_container'
+import { Testing } from '@defichain/jellyfish-testing'
+import BigNumber from 'bignumber.js'
+
+describe('Loan listVault', () => {
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/loan/listVaults.test.ts:5]:
```suggestion
describe('Loan listVaults', () => {
```

File: packages/jellyfish-api-core/__tests__/category/loan/listVaults.test.ts [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    await testing.rpc.oracle.setOracleData(oracleId, timestamp, { prices: [{ tokenAmount: '2@TSLA', currency: 'USD' }] })
+    await testing.rpc.oracle.setOracleData(oracleId, timestamp, { prices: [{ tokenAmount: '2@AAPL', currency: 'USD' }] })
+
+    // collateral tokens
+    await testing.rpc.loan.setCollateralToken({
+      token: 'DFI',
+      factor: new BigNumber(1),
+      priceFeedId: 'DFI/USD'
+    })
+
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/loan/listVaults.test.ts:41]: I think you purposely skip `await testing.generate(1) `here

File: packages/jellyfish-api-core/__tests__/category/loan/listVaults.test.ts [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+    await testing.generate(1)
+    // make vault enter under liquidation state by a price hike of the loan token
+    const timestamp = Math.floor(new Date().getTime() / 1000)
+    await testing.rpc.oracle.setOracleData(oracleId, timestamp, { prices: [{ tokenAmount: '1000@AAPL', currency: 'USD' }] })
+    await testing.generate(1)
+
+    // List vaults
+    // should be only vaultId1, vaultId2, vaultId3 coz isUnderLiquidation filter default value equals to false
+    const vaults = await testing.rpc.loan.listVaults()
+    expect(vaults).toStrictEqual(expect.arrayContaining([
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/loan/listVaults.test.ts:107]:
```suggestion
    expect(vaults).toStrictEqual([
```

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/loan/listVaults.test.ts:107]: You do not need `expect.arrayContaining`

File: packages/jellyfish-api-core/__tests__/category/loan/listVaults.test.ts [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+  })
+
+  it('should listVaults with pagination', async () => {
+    // List vaults
+    const vaults = await testing.rpc.loan.listVaults({ limit: 2 })
+    expect(Object.keys(vaults).length).toStrictEqual(2)
+
+    // fetch the second page
+    const vaultsSecondPage = await testing.rpc.loan.listVaults({ including_start: false, start: vaults[Object.keys(vaults).length - 1].vaultId })
+    // should be 2 entries
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/loan/listVaults.test.ts:298]:
```suggestion
    // should be 1 entry
```

File: docs/node/RPC Category/16-loan.md [hunk 9]
```diff
@@ -330,6 +330,48 @@ interface loan {
...
+}
+```
+
+## listVaults
+
+List all available vaults.
+
+```ts title="client.loan.listVaults()"
+interface loan {
+  listVaults (options: ListVaultOptions, pagination: VaultPagination): Promise<VaultDetails[]>
```
Comment (jingyi2811) [inline docs/node/RPC Category/16-loan.md:358]: listVaults (pagination: VaultPagination = {}, options: ListVaultOptions = {}): Promise<VaultDetails[]>

Like what you wrote in the js file

File: docs/node/RPC Category/16-loan.md [hunk 10]
```diff
@@ -330,6 +330,48 @@ interface loan {
...
+}
+
+interface ListVaultOptions {
+  ownerAddress?: string
+  loanSchemeId?: string
+  isUnderLiquidation?: boolean
+}
+
+interface VaultPagination {
+  start?: number
```
Comment (jingyi2811) [inline docs/node/RPC Category/16-loan.md:368]: This should be string
Like what you wrote in the js file

**Review by surangap — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**